Added DisplayName to ParameterSymbol and choices

bozturkMSFT opened the pull request:
### Problem
<!-- Add the issue number if exists. Describe the problem otherwise. -->
Fixes https://github.com/dotnet/templating/issues/2822

### Solution
<!-- Describe the solution. -->
Added `DisplayName` property to `ParameterSymbol`, `ICacheParameter` and `ICacheTag`.

### Checks:
- [x] Added unit tests
- [x] Added `#nullable enable` to all the modified files [?](https://github.com/dotnet/templating/wiki/Contributing#coding-style)

vlada-shubina commented on src/Microsoft.TemplateEngine.Abstractions/ICacheParameter.cs:
```diff
@@ -1,11 +1,33 @@
 
-        string DefaultValue { get; }
+        /// <summary>
+        /// Gets the friendly name of the parameter to be displayed to the user.
```
Probably worth to mention that DisplayName and Description are localizable but not sure

vlada-shubina commented on src/Microsoft.TemplateEngine.Abstractions/ICacheParameter.cs:
```diff
@@ -1,11 +1,33 @@
 namespace Microsoft.TemplateEngine.Abstractions
 {
+    /// <summary>
+    /// Defines the representation of a parameter in template cache.
```
I think worth mentioning the `choice` parameters are not included

vlada-shubina commented on src/Microsoft.TemplateEngine.Abstractions/ICacheTag.cs:
```diff
@@ -1,13 +1,29 @@
 namespace Microsoft.TemplateEngine.Abstractions
 {
+    /// <summary>
+    /// Represents a parameter symbol with choices in template cache file.
```
This also represents `tags` from `template.json`.

vlada-shubina commented on src/Microsoft.TemplateEngine.Abstractions/ICacheTag.cs:
```diff
@@ -1,13 +1,29 @@
     {
-        string Description { get; }
+        /// <summary>
+        /// Gets the friendly name of the tag to be displayed to the user.
```
Using `tag` is misleading, probably better `choice parameter symbol`.

vlada-shubina commented on src/Microsoft.TemplateEngine.Abstractions/IParameterSymbolLocalizationModel.cs:
```diff
@@ -1,13 +1,23 @@
+        string? Description { get; }
 
-        IReadOnlyDictionary<string, string> ChoicesAndDescriptions { get; }
+        IReadOnlyDictionary<string, ParameterChoiceLocalizationModel> Choices { get; }
```
is it safe change? Can it make older versions of `TemplateCache` unreadable? (meaning renaming `ChoicesAndDescriptions`)

DavidKarlas replied:
Version -> Hive -> cache.json file will be different, going forward with Discovery this becomes even safer change, since Cache is versioned and thrown away if cache version is out of date.

vlada-shubina replied:
Lately I realized that we use same class for blob storage cache file, so we cannot change it without adding new version for search cache. 
I think we can change it still if we keep serialization same as before. We can also do this change as we work on improvements for search cache - we will need new version anyway.

bozturkMSFT replied:
I'm keeping this change since we will be changing search cache file in .NET 6 anyway. Any concerns?

bozturkMSFT replied:
Created the tag [TemplateDiscovery-CacheVersion-1-0](https://github.com/dotnet/templating/releases/tag/TemplateDiscovery-CacheVersion-1-0) from which we can create a branch in the future if needed.

How can we prevent creating a new branch like this every time we make a change to cache file format? Hopefully, once we change the nuget search file to a custom format, changing the cache format won't cause this problem. Is that right?

vlada-shubina replied:
I think we should think carefully of new format and allow only additive changes. If we need modifications, we will need versioning approach.

vlada-shubina commented on src/Microsoft.TemplateEngine.Abstractions/ICacheTag.cs:
```diff
@@ -1,13 +1,29 @@
+        string? Description { get; }
 
-        IReadOnlyDictionary<string, string> ChoicesAndDescriptions { get; }
+        IReadOnlyDictionary<string, ParameterChoice> Choices { get; }
```
Visual Studio and  Mono uses this interface, I'm not sure about changing it

http://index/?leftProject=Microsoft.TemplateEngine.Abstractions&leftSymbol=kzkyqi5zfqek

vlada-shubina replied:
Actually just realized further impact - `TemplateInfo` is also used in Template Discovery tool. If we rename it now, the file generated by the tool won't be compatible with previous versions.

bozturkMSFT replied:
Since we moved this work to .NET 6,  can we keep this change?

vlada-shubina commented on src/Microsoft.TemplateEngine.Orchestrator.RunnableProjects/Localization/ParameterSymbolLocalizationModel.cs:
```diff
@@ -1,21 +1,29 @@
+            DisplayName = displayName;
             Description = description;
-            ChoicesAndDescriptions = choicesAndDescriptions;
+            Choices = choicesAndDescriptions;
```
If rename is kept, please rename the parameter for consistency

vlada-shubina reviewed (commented):
Not sure about renames in Abstractions interfaces - those members are used by VS and Mono

vlada-shubina commented on test/Microsoft.TemplateEngine.Cli.UnitTests/CacheUpgradeTests.cs:
```diff
@@ -49,8 +49,8 @@ private static JObject CacheDataVersion1000
       ""Tags"": {
                 ""type"": {
                     ""Description"": null,
-          ""ChoicesAndDescriptions"": {
```
I think this change actually breaks the test, it seems the idea of the test is that the code can still read older version of the cache.

bozturkMSFT replied:
I see. What should be the proper action to take here, then? Revert the changes to this JSON, but also update `TemplateInfoReaderVersion1_0_0_0` to still be able to read this old format?

vlada-shubina replied:
It depends whether we are keeping support of reading old versions of cache or not. 

1) If we do keep support: we need to create another version (1.0.0.4) to read new format and keep old versions intact.
2) If we don't want to keep support: remove all the versions and keep implementation which supports reading just latest version of cache.

vlada-shubina reviewed (changes requested):
Modifying `ChoicesAndDescriptions` is not safe and will need further changes.

DavidKarlas commented:
When I was reviewing this PR I was also conflicted about if change from `ChoicesAndDescriptions` to `Choices` is OK and worth it...
But I had 2 conclusions:
a) Value of dictionary is changing from `string` to `ParameterChoice` which is as big breaking change as property name
b) It effects mostly just cache which, is with every new version of host thrown away...

bozturkMSFT commented:
Removing template cache versioning is going to be more complicated than it looks and I don't want to add that kind of complexity to this PR. Therefore, I just added 1_0_0_4 cache serializer and created a ticket for removing them here https://github.com/dotnet/templating/issues/2929.

Please see the last commit only. The changes include:
- Reverting the changes to `TemplateInfoReaderVersion1_0_0_0`, `TemplateInfoReaderInitial` and cache upgrade tests.
- Added TemplateInfoReaderVersion1_0_0_4 and updated the settings cache version number to 1_0_0_4.

DavidKarlas commented:
I'm not sure adding 1.0.0.4 version and reader was worth the effort... But I think this is good to merge.
